jscomp/tools/ppx_class.ml

```diff
@@ -0,0 +1,207 @@
+*)
+
+(* Author: Hongbo Zhang  *)
+open Ast_mapper
+open Ast_helper
+open Asttypes
+open Parsetree
+open Longident
+
+
+(* Define types of indications supported by the FFI *)
+type indication = 
+  | Indication_read
+  | Indication_write
+  | Indication_unsafe
+
+(* Define a comparator to sort indications with *)
+let compare_indication : indication -> indication -> int = compare
+
+(* Given an attribute string, determine the indications *)
+let classify_indication name = match name with
+  | "r" -> [Indication_read]
+  | "w" -> [Indication_write]
+  | "rw" -> [Indication_read;Indication_write]
+  | "unsafe" -> [Indication_unsafe]
+  | _ -> []
+
+(* Get unique indications from an attribute list *)
+let indications_of_attr attrlst = 
+  List.sort_uniq compare_indication (List.concat
+    (List.map (fun ({txt=x;_},_) -> classify_indication x) attrlst))
+
+(* Define the methods generated by this ppx extension *)
+type methods = 
+  | Method_read
+  | Method_read_unsafe
+  | Method_write
+  | Method_arity
+  | Method_arity_unsafe
+
+(* Define the formatter for printing errors *)
+let fmtr = Format.err_formatter
+
+
+(* Predicated print warning to stderr *)
+let pwarn loc predicate message = 
+  if predicate then 
+    begin
+      let open Location in
+      let () = Printf.fprintf stderr "\n%!" in
+      let () = print_loc fmtr loc in
+      Format.fprintf fmtr ": %s\n%!" message
+    end
+  else 
+    ()
+
+(* Define which methods get generated based on indications *)
+let methods_of_indications i loc = 
+  match i with
+  | [] -> [Method_arity]
+  | [Indication_read] -> [Method_read]
+  | [Indication_read;
+     Indication_unsafe] -> [Method_read_unsafe]
+  | [Indication_read;
+     Indication_write] -> [Method_read;Method_write]
+  | [Indication_read;
+     Indication_write;
+     Indication_unsafe] -> [Method_read_unsafe;Method_write]
+  | [Indication_write] -> [Method_write]
+  | [Indication_unsafe] -> [Method_arity_unsafe]
+  | _ -> begin
+      pwarn loc true "Invalid combination of indications."; 
+      []
+    end
+
+(* Filter out the attributes expanded by this ppx from an attribute list *)
+let filter_attributes attrlst = List.filter 
+    (fun ({txt=x;_},_) -> classify_indication x = []) attrlst
+
+(* Filter out the attributes expanded by this ppx from a core_type*)
+let strip_attr t = {t with ptyp_attributes=filter_attributes t.ptyp_attributes}
+
+(* Count the arity of a core_type *)
+let rec count_arity z = match z with
+  | {ptyp_desc = Ptyp_arrow (_,_,y);_} -> 1 + (count_arity y)
+  | _ -> 0
+
+(* Determine whether a core_type is a non-nullary function *)
+let is_non_nullary_fn z = (count_arity z) > 0
+
+(* Produce a new core_type from the given core_type, removing 'opt/option/def'
+   from the return type*)
+let rec optless typ loc prefix = match typ with
+  | {ptyp_desc = Ptyp_constr ({txt=(Lident ("opt"|"def"|"option")) | 
+      (Ldot (Lident "Js",("opt"|"def"|"option")));_},[x]);_} -> x
```

Comment:
there's no JS.option right? also, I thought you mentioned there's another constructor `null`?